adr/ADR-56.md

```diff
@@ -50,3 +51,100 @@ The interactions between `PersistMode:async` and `sync:always` are as follows:
+| `AllowDirect` disabled.                  | `$JS.API.STREAM.MSG.GET.<stream>` | An API meant only for management outside of the hot path. The request goes to every server, but is normally only answered by the stream leader. | Current highest level of read consistency. Only the stream leader answers, but stale reads are technically possible after leader changes or during network partitions since an old leader could still answer before the current leader does. |
+| `AllowDirect` enabled.                   | `$JS.API.DIRECT.GET.<stream>`     | If the stream is replicated, the followers will also answer read requests.                                                                      | Higher availability read responses but with lower consistency. A read request will be randomly served by a server hosting the stream. Recently written data is not guaranteed to be returned on a subsequent read request.                   |
+| `MirrorDirect` enabled on mirror stream. | `$JS.API.DIRECT.GET.<stream>`     | If the stream is mirrored, the mirror can also answer read requests. For example a mirror stream in a different cluster or on a leaf node.      | Higher availability with potential of fast local read responses but with lowest consistency. Mirrors can be in any relative state to the source.                                                                                             |
+
+Additionally, if a stream is replicated and a consumer is created, there is no guarantee that the consumer can
+immediately observe all the written messages at that time. For example, if a R1 consumer is created on a follower not up
+to date on all writes yet. The consumer will eventually observe all the writes as it keeps on fetching new messages as
+they come in.
+
+Newer server versions, like for 2.14+, should support more configurability or in general higher levels of consistency as
+opt-in. For example, higher read consistency for consumers can be achieved by having consumer CRUD operations go through
+the stream's Raft log instead of the Meta Raft log, which ensures that a consumer created at time X in the stream log
+can observe all the stream writes up to time X.
+
+Specifically, higher level consistency for message read requests would roughly require:
+
+- `AllowDirect` should not need to be disabled. The `$JS.API.STREAM.MSG.GET.<stream>` API, when `AllowDirect` is
+  disabled, has significant overhead since these requests go to ALL servers not just the servers hosting the stream.
+- Direct Get allows using batch requests, this should also be supported. (Which is not the case with the Msg Get API
+  above)
+
+Linearizable reads would be desirable, but a minimum would be to opt in to at least session-level guarantees such as
+reading your own writes and monotonic reads.
+
+### Discussion
+
+#### What would be expected given different topologies?
```

Comment:
Starting a thread for thoughts around this topic:

Location transparency is a big topic for NATS. With that in mind "it should not matter" where you are connected in the topology. If you need a high consistency read, you shouldn't need to care where you are connected and if the stream happens to be in the same place.

I'd say that a client connected to a leaf node, connected to a cluster, gateway-ed to another cluster that hosts the stream.. should also be able to not only write to that stream, but also get a consistent read when it requires it. We shouldn't require the client to be connected directly to the cluster hosting the stream.